src/bm_sim/lpm_trie.cpp

```diff
@@ -0,0 +1,278 @@
+/* Copyright 2013-present Barefoot Networks, Inc.
+ * Copyright 2021 VMware, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <bm/bm_sim/_assert.h>
+#include <lpm_trie.h>
+
+#include <unordered_map>
+#include <map>
+namespace bm {
+
+struct Branch {
+  byte_t value;
+  std::unique_ptr<Node> next;
+};
+
+class Node {
+ public:
+  explicit Node(Node *parent = nullptr, byte_t child_id = 0)
+      : parent(parent), child_id(child_id) {}
+  Node(const Node &) = delete;
+
+  Node *get_next_node(byte_t byte) const;
+
+  void set_next_node(byte_t byte, std::unique_ptr<Node> next_node) {
+    add_branch(byte, std::move(next_node));
+  }
+
+  Node *get_parent() const { return parent; }
+
+  byte_t get_child_id() const { return child_id; }
+
+  // Prefix
+
+  bool get_prefix(uint8_t prefix_length, byte_t key, value_t *value);
+
+  void insert_prefix(uint8_t prefix_length, byte_t key, value_t value);
+
+  bool delete_prefix(uint8_t prefix_length, byte_t key);
+
+  // Branch
+  void add_branch(byte_t byte, std::unique_ptr<Node> nextNode);
+
+  bool delete_branch(byte_t byte);
+
+  bool is_empty() const { return prefixes.empty() && branches.empty(); }
+
+ private:
+  std::map<std::pair<uint8_t, byte_t>, value_t> prefixes;
```

Comment:
this feels different from the earlier version I reviewed. Did you switch from a `vector` to a `map`? A sorted vector should be faster to search (and more compact) than a map, so why make this change?